src/metabase/public_settings/premium_features.clj

```diff
@@ -240,3 +242,179 @@
+  availability of EE code and the necessary premium feature. Returns a fn which, when invoked, applies its args to one
+  of the EE implementation, the OSS implementation, or the fallback function."
+  [ee-ns ee-fn-name]
+  (fn [& args]
+    (u/ignore-exceptions (classloader/require ee-ns))
+    (let [{:keys [ee oss feature fallback]} (get @registry ee-fn-name)]
+      (cond
+        (and ee (check-feature feature))
+        (apply ee args)
+
+        (and ee (fn? fallback))
+        (apply fallback args)
+
+        :else
+        (apply oss args)))))
+
+(defn- validate-ee-args
+  "Throws an exception if the required :feature option is not present."
+  [{feature :feature :as options}]
+  (when-not feature
+    (throw (ex-info (trs "The :feature option is required when using defenterprise in an EE namespace!")
+                    {:options options}))))
+
+(defn- oss-options-error
+  "The exception to throw when the provided option is not included in the `options` map."
+  [option options]
+  (ex-info (trs "{0} option for defenterprise should not be set in an OSS namespace! Set it on the EE function instead." option)
+           {:options options}))
+
+(defn validate-oss-args
+  "Throws exceptions if EE options are provided, or if an EE namespace is not provided."
+  [ee-ns {:keys [feature fallback] :as options}]
+  (when-not ee-ns
+    (throw (Exception. (str (trs "An EE namespace must be provided when using defenterprise in an OSS namespace!")
+                            " "
+                            (trs "Add it immediately before the argument list.")))))
+  (when feature (throw (oss-options-error :feature options)))
+  (when fallback (throw (oss-options-error :fallback options))))
+
+(defn- docstr-exception
+  "The exception to throw when defenterprise is used without a docstring."
+  [fn-name]
+  (Exception. (tru "Enterprise function {0}/{1} does not have a docstring. Go add one!" (ns-name *ns*) fn-name)))
+
+(defmacro defenterprise-impl
+  "Impl macro for `defenterprise` and `defenterprise-schema`. Don't use this directly."
+  [{:keys [fn-name docstr ee-ns fn-tail options schema? return-schema]}]
+  (when-not docstr (throw (docstr-exception fn-name)))
+  (let [oss-or-ee (if (in-ee?) :ee :oss)]
+    (case oss-or-ee
+      :ee  (validate-ee-args options)
+      :oss (validate-oss-args '~ee-ns options))
+    `(let [ee-ns#        '~(or ee-ns (ns-name *ns*))
+           ee-fn-name#   (symbol (str ee-ns# "/" '~fn-name))
+           oss-or-ee-fn# ~(if schema?
+                           `(schema/fn ~(symbol (str fn-name)) :- ~return-schema ~@fn-tail)
+                           `(fn ~(symbol (str fn-name)) ~@fn-tail))]
+        (register-mapping! ee-fn-name# (merge ~options {~oss-or-ee oss-or-ee-fn#}))
+        (def
+          ~(vary-meta fn-name assoc :arglists ''([& args]))
+          (dynamic-ee-oss-fn ee-ns# ee-fn-name#)))))
+
+(defn- options-conformer
+  [conformed-options]
+  (into {}
+        (map #(-> % second vals vec)
```

Comment:
we're dealing with small maps so the order of `vals` is consistent but this isn't super well-defined. `{:k :feature :v :ee-token}` -> `[:feature :ee-token]` is what is going on here. Relying on the order of vals is always a bit weird to me. maybe replace with `(map (comp (juxt :k :v) second) conformed-options)`

Reply:
Good call, will update